quickcheck/Test/Tasty/QuickCheck.hs

```diff
@@ -259,14 +263,21 @@ quickCheck :: (Progress -> IO ())
            -> QC.Args
            -> QC.Property
            -> IO QC.Result
-quickCheck yieldProgress args prop = do
+quickCheck yieldProgress args@(QC.Args {QC.maxSuccess = argMaxSuccess}) prop = do
   -- Here we rely on the fact that QuickCheck currently prints its progress to
   -- stderr and the overall status (which we don't need) to stdout
+  maxSuccess <- newIORef $ fromIntegral argMaxSuccess
+  let setMaxTests r = case maybeNumTests r of
+        Just n -> unsafePerformIO $ writeIORef maxSuccess (fromIntegral n) $> r
+        Nothing -> r
   tm <- QC.newTerminal
           (const $ pure ())
-          (\progressText -> yieldProgress emptyProgress { progressPercent = parseProgress progressText })
+          $ \progressText -> do
+            ms <- readIORef maxSuccess
+            yieldProgress emptyProgress { progressPercent = 100 * parseProgress progressText / ms }
```

Comment:
Assuming `parseProgress` returns the number of completed tests this code is "fine" as the 100 is just there to convert from fraction to percent.

The bigger issue here is that you can also set the number of tests with `withMaxSuccess` and this code doesn't capture that. However, it's a bit tricky to deal with that given the current interface to QuickCheck.

Reply:
`parseProgress` does just returns the number of tests run.

"this code doesn't capture that" do you mean the current implementation or the changes that I have made?

Reply:
I don't think neither the changes you've made or the original code captures the fact that `withMaxSuccess` changes the `Result` type to change the number of tests. The best way I see for you to account for that change is to use a `callback` in QuickCheck and update an `IORef` when you see the `maybeMaxSuccess` field change in the `Result` argument to your callback.

Reply:
I have tried this and the changes I have made do indeed fix the problem, but this may not be the optimum solution.

I am unfamiliar with `callback` could you give me a link describing how to use it?

Reply:
Ok, if this works then there is no point doing it differently.

Reply:
Does the callback remove the need for unsafePerformIO?